src/outpack/location_pull.py

```diff
@@ -0,0 +1,465 @@
+                )
+                _location_pull_files_archive(packet.packet, store, root)
+
+            mark_known(
+                root, packet.packet, LOCATION_LOCAL, packet.hash, time.time()
+            )
+
+    print(
+        f"Unpacked {n_packets} {pl(n_packets, 'packet')} in "
+        f"{humanize.time.precisedelta(int(time.time() - time_start))}."
+    )
+
+    return list(plan.packets.keys())
+
+
+# This approach may be suboptimal in the case where the user does not
+# already have a file store, as it means that files will be copied
+# around and hashed more than ideal:
+#
+# * hash the candidate file
+# * rehash on entry into the file store
+# * copy into the file store
+# * copy from the file store into the final location
+#
+# So in the case where a hash is only present once in a chain of
+# packets being pulled this will be one too many hashes and one too
+# many copies.
+#
+# However, this approach makes the logic fairly easy to deal with,
+# and copes well with data races and corruption of data on disk
+# (e.g., users having edited files that we rely on, or editing them
+# after we hash them the first time).
+@contextmanager
+def _location_pull_files(
+    files: List[PacketFileWithLocation], root: OutpackRoot
+) -> Generator[FileStore, None, None]:
+    store = root.files
+    cleanup_store = False
+    if store is not None:
+        exists, missing = partition(lambda file: store.exists(file.hash), files)
+
+        if exists:
+            print(
+                f"Found {len(exists)} {pl(exists, 'file')} in the "
+                f"file store"
+            )
+    else:
+        print("Looking for suitable files already on disk")
+        store = _temporary_filestore(root)
+        cleanup_store = True
+
+        missing = []
+        no_found = 0
+        for file in files:
+            path = find_file_by_hash(root, file.hash)
+            if path is not None:
+                store.put(path, file.hash)
+                no_found += 1
+            else:
+                missing.append(file)
+
+        print(f"Found {no_found} {pl(no_found, 'file')} on disk ")
+
+    if len(missing) == 0:
+        print("All files available locally, no need to fetch any")
+    else:
+        locations = {file.location for file in missing}
+        total_size = humanize.naturalsize(sum(file.size for file in missing))
+        print(
+            f"Need to fetch {len(missing)} {pl(missing, 'file')} "
+            f"({total_size}) from {len(locations)} "
+            f"{pl(locations, 'location')}"
+        )
+        for location in locations:
+            from_this_location = [
+                file for file in missing if file.location == location
+            ]
+            _location_pull_hash_store(
+                from_this_location,
+                location,
+                _location_driver(location, root),
+                store,
+            )
+
+    yield store
+    if cleanup_store:
+        store.destroy()
```

Comment:
```suggestion
    try:
        yield store
    finally:
        if cleanup_store:
            store.destroy()
```

in case the yield (ie. the contents of the `with`-block) raises an exception.